test/typed_record_test.exs

```diff
@@ -0,0 +1,252 @@
+
+        @type user() :: {:user, integer, String.t(), String.t(), atom}
+      end
+
+    type0 = extract_first_type(bytecode1)
+    type1 = extract_first_type(@bytecode1)
+
+    assert type0 == type1
+
+    # Define a second struct with the type expected for TestStruct.
+    {:module, _name, bytecode2, _exports} =
+      defmodule TestRecordTag do
+        require Record
+        Record.defrecord(:user, User, one: 1, two: nil, three: "def", four: :hi)
+
+        @type user() :: {User, integer, String.t(), String.t(), atom}
+      end
+
+    type1 = extract_first_type(bytecode2)
+    type2 = extract_first_type(@bytecode2)
+
+    assert type1 == type2
+
+    type3 = extract_first_type(@bytecode3)
+
+    assert type0 == type3
+  end
+
+  test "generates an opaque type if `opaque: true` or `visibility: :opaque` is set" do
+    # Define a second struct with the type expected for TestStruct.
+    {:module, _name, bytecode_expected, _exports} =
+      defmodule TestRecord2 do
+        require Record
+        Record.defrecord(:user, int: 1)
+
+        @opaque user() :: {:user, integer}
+      end
+
+    type1 = extract_first_type(@bytecode_opaque1, :opaque)
+    type2 = extract_first_type(bytecode_expected, :opaque)
+
+    assert type1 == type2
+
+    type3 = extract_first_type(@bytecode_opaque2, :opaque)
+
+    assert type3 == type2
+  end
+
+  test "generates a private type if `visibility: private` is set" do
+    # Define a second struct with the type expected for TestStruct.
+    {:module, _name, bytecode_private_expected, _exports} =
+      defmodule TestRecord3 do
+        require Record
+        Record.defrecord(:user, int: 1)
+
+        @typep user() :: {:user, integer}
+        @opaque t2 :: user
+      end
+
+    type1 = extract_first_type(@bytecode_private, :typep)
+    type2 = extract_first_type(bytecode_private_expected, :typep)
+
+    assert type1 == type2
+  end
+
+  test "generates the struct in a submodule if `module: ModuleName` is set" do
+    require TestRecModule.Rec
+    assert TestRecModule.Rec.user() == {:user, 1}
+  end
+
+  ############################################################################
+  ##                                Problems                                ##
+  ############################################################################
+
+  test "Typedrecord missing record name" do
+    assert_raise CompileError, ~r"undefined function typedrecord/1", fn ->
+      defmodule ScopeTest do
+        use TypedStruct
+
+        typedrecord do
+          field :in_scope, term()
+        end
+      end
+    end
+  end
+
+  test "it is not possible to add twice a field with the same name" do
+    assert_raise ArgumentError, "the field :name is already set", fn ->
+      defmodule InvalidStruct do
+        use TypedStruct
+
+        typedrecord :user do
+          field :name, String.t()
+          field :name, integer()
+        end
+      end
+    end
+  end
+
```

Comment:
There are other tests in the test plan for `typedstruct` that would be interesting to have as well here:

* “TypedStruct macros are available only in the typedrecord block”
* “the name of a field must be an atom” (is it true for records as well? If yes, this test should be there)
* “aliases are properly resolved in types”